store_test.go

```diff
@@ -0,0 +1,249 @@
+	intVal := 1
+	tests := []struct {
+		store        store
+		wantExistKey []string
+	}{
+		{
+			store{},
+			[]string{"env", "vars", "steps"},
+		},
+		{
+			store{
+				steps: []map[string]any{},
+				vars: map[string]any{
+					"key": "value",
+				},
+			},
+			[]string{"env", "vars", "steps"},
+		},
+		{
+			store{
+				useMap: true,
+			},
+			[]string{"env", "vars", "steps"},
+		},
+		{
+			store{
+				parentVars: map[string]any{
+					"key": "value",
+				},
+			},
+			[]string{"env", "vars", "steps"},
+		},
+		{
+			store{
+				bindVars: map[string]any{
+					"bind": "value",
+				},
+			},
+			[]string{"env", "vars", "steps", "bind"},
+		},
+		{
+			store{
+				loopIndex: &intVal,
+			},
+			[]string{"env", "vars", "steps", "i"},
+		},
+		{
+			store{
+				cookies: map[string]map[string]*http.Cookie{},
+			},
+			[]string{"env", "vars", "steps", "cookies"},
+		},
+	}
+	trns := cmp.Transformer("Sort", func(in []string) []string {
+		out := append([]string(nil), in...) // Copy input to avoid mutating it
+		sort.Strings(out)
+		return out
+	})
+	for _, tt := range tests {
+		got := tt.store.toNormalizedMap()
+		gotKeys := make([]string, 0, len(got))
+		for k := range got {
+			gotKeys = append(gotKeys, k)
+		}
+		if diff := cmp.Diff(gotKeys, tt.wantExistKey, trns); diff != "" {
+			t.Errorf("%s", diff)
+		}
+	}
+}
+
+func TestRecordToCookie(t *testing.T) {
+	cookie1 := http.Cookie{
+		Name:   "key1",
+		Value:  "value1",
+		Domain: "example.com",
+	}
+	cookie2 := http.Cookie{
+		Name:   "key2",
+		Value:  "value2",
+		Domain: "example.com",
+	}
+	cookie3 := http.Cookie{
+		Name:   "key3",
+		Value:  "value3",
+		Domain: "sub.example.com",
+	}
+	cookie4 := http.Cookie{
+		Name:   "key1",
+		Value:  "value4",
+		Domain: "example.com",
+	}
+	cookie5 := http.Cookie{
+		Name:    "key1",
+		Value:   "value4",
+		Domain:  "example.com",
+		Expires: time.Now(),
+	}
+	tests := []struct {
+		store   store
+		cookies []*http.Cookie
+		want    map[string]map[string]*http.Cookie
+	}{
+		{
+			store{},
+			[]*http.Cookie{},
+			map[string]map[string]*http.Cookie{},
+		},
+		{
+			store{},
+			[]*http.Cookie{&cookie1},
+			map[string]map[string]*http.Cookie{
+				"example.com": {
+					"key1": &cookie1,
+				},
+			},
+		},
+		{
+			store{},
+			[]*http.Cookie{&cookie1, &cookie2},
+			map[string]map[string]*http.Cookie{
+				"example.com": {
+					"key1": &cookie1,
+					"key2": &cookie2,
+				},
+			},
+		},
+		{
+			store{},
+			[]*http.Cookie{&cookie1, &cookie2, &cookie3},
+			map[string]map[string]*http.Cookie{
+				"example.com": {
+					"key1": &cookie1,
+					"key2": &cookie2,
+				},
+				"sub.example.com": {
+					"key3": &cookie3,
+				},
+			},
+		},
+		{
+			store{
+				cookies: map[string]map[string]*http.Cookie{
+					"example.com": {
+						// Override
+						"key1": &cookie4,
+					},
+				},
+			},
+			[]*http.Cookie{&cookie5},
+			map[string]map[string]*http.Cookie{
+				// Expire
+				"example.com": {},
```

Comment:
Once all cookies are deleted, should I also delete the domain key itself?

Reply:
I have no strong opinion on this point.

So I tried it in Chrome.

I erased all cookies, but the domain on the display remained.
<img width="446" alt="CleanShot 2023-07-17 at 08 11 01@2x" src="https://github.com/k1LoW/runn/assets/57114/c754c0b3-a2ca-4c61-99f5-85fa181f5a1f">

Reply:
Thank you for looking into this!
Gives me more confidence in these specs.